### ISSUE TITLE
How to read DTC codes ?

### ISSUE THREAD

**Issue by thanhhai16** [mickeyl/LTSupportAutomotive#14]

**Comment by thdankert** [comment]
If you already have a connection to your car, you can get the DTCs by sending the appropriate PIDs and parsing the response.

* LTOBD2PID_PERMANENT_DTC_0A
* LTOBD2PID_PENDING_DTC_07
* LTOBD2PID_STORED_DTC_03

The exact command depends on the kind of DTC you want to read, as there may be some only relevant to a freeze frame, or some are stored permanently due to repeated errors.

Either way, if there are DTC reported, the response will have an array `troubleCodes` which contains `LTOBD2DTC`. These have a `code` and an `explanation`.

If the explanation is empty, it's probably a manufacturer-defined code: These are usually explained in your car maker's service/workshop manual.

**Comment by janessbach** [comment]
Is there also the possibility to delete the error codes?

**Comment by mickeyl** [comment]
Yes, sending ´LTOBD2PID_CLEAR_STORED_DTC_04´ will reset the MIL and erase the stored DTCs.

**Comment by janessbach** [comment]
If I am using the following code as example, I am not able to read any error codes from the ecu's. 

`import Foundation

struct ErrorDescription: Codable {
    var errorCode: String
    var description: String
}

class BluetoothOBD {
    
    private var serviceIds: [String] = ["FFF0", "FFE0", "BEEF", "E7810A71-73AE-499D-8C15-FAA9AEF0C3F2"]
    private var transporter: LTBTLESerialTransporter?
    private var adapter: LTOBD2AdapterELM327?
    
    func start() -> Void {
        
        self.transporter = LTBTLESerialTransporter.init(identifier: nil, serviceUUIDs: serviceIds.map({ (id) -> CBUUID in return CBUUID(string: id) }))
        
        self.transporter!.connect { (input, output) in
            if (input == nil) {
                print("Could not connect to OBD2 adapter");
                return;
            }
            self.adapter = LTOBD2AdapterELM327.init(inputStream: input!, outputStream: output!)
            self.adapter!.connect()
        }
        
        self.transporter?.startUpdatingSignalStrength(withInterval: 1.0)
        
        NotificationCenter.default.addObserver(self, selector: #selector(onAdapterChangedState(_:)), name: Notification.Name(LTOBD2AdapterDidUpdateState), object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(onTransporterDidUpdateSignalStrength(_:)), name: Notification.Name(LTBTLESerialTransporterDidUpdateSignalStrength), object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(onAdapterDidSendBytes(_:)), name: Notification.Name(LTOBD2AdapterDidSend), object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(onAdapterDidReceiveBytes(_:)), name: Notification.Name(LTOBD2AdapterDidReceive), object: nil)
    }
    
    func requestVIN(handler: @escaping(String) -> Void) {
        let vin = LTOBD2PID_VIN_CODE_0902.init()
        
        self.adapter?.transmitMultipleCommands([vin], responseHandler: { (command) in
            handler(vin.formattedResponse)
        })
    }
    
    func requestError(handler: @escaping([ErrorDescription]) -> Void) {
        let permanent: LTOBD2PID_PERMANENT_DTC_0A = LTOBD2PID_PERMANENT_DTC_0A.init()
        let pending: LTOBD2PID_PENDING_DTC_07 = LTOBD2PID_PENDING_DTC_07.init()
        let stored: LTOBD2PID_STORED_DTC_03 = LTOBD2PID_STORED_DTC_03.init()
        
        self.adapter?.transmitMultipleCommands([permanent, pending, stored], responseHandler: { (command) in
            
            var errors = self.handleErrorCode(errorCodes: permanent.troubleCodes)
            errors += self.handleErrorCode(errorCodes: pending.troubleCodes)
            errors += self.handleErrorCode(errorCodes: stored.troubleCodes)
            
            handler(errors)
        })
    }
    
    private func handleErrorCode(errorCodes: [LTOBD2DTC]) -> [ErrorDescription] {
        return errorCodes.map { (code) -> ErrorDescription in
            return ErrorDescription(errorCode: code.code, description: code.description)
        }
    }
    
    @objc func onAdapterChangedState(_ notification:Notification) {
        print("onAdapterChangedState \(notification)")
    }
    
    @objc func onTransporterDidUpdateSignalStrength(_ notification:Notification) {
        print("onTransporterDidUpdateSignalStrength \(notification)")
    }
    
    @objc func onAdapterDidSendBytes(_ notification:Notification) {
        print("onAdapterDidSendBytes \(notification)")
    }
    
    @objc func onAdapterDidReceiveBytes(_ notification:Notification) {
        print("onAdapterDidReceiveBytes \(notification)")
    }
    
}
`

I have confirmed that other iOS apps are able to read DTCs, so there are some stored. What I have seen, is that the response of the request(s) are handled nearly immediately... in the other apps there is a progress bar and the process takes about 10 seconds. So I think I miss something here. 

Any ideas?
As setup I am using an VGate iCar Pro BLE4.0 Dual Adapter and using the Swift ObjectiveC brideing mode. I am able to read the VIN so the connection works.

**Comment by mukulpandey57** [comment]
Hello @mickeyl , @thdankert 
I am getting incorrect DTC in my code using this Library.
From Other App i am getting "P0441","P0442","P0446" and in app i am getting "C0104" "C0204"
Please help it's urgent.
![Screenshot 2020-08-24 at 2 22 21 PM](https://user-images.githubusercontent.com/8499425/91024401-877c6680-e615-11ea-9055-0db301f294cf.png)
![Screenshot 2020-08-24 at 2 23 23 PM](https://user-images.githubusercontent.com/8499425/91024442-9531ec00-e615-11ea-8206-3a35dbc6218e.png)

**Comment by thdankert** [comment]
@mukulpandey57 Is your issue related to this one? The OP asked for general information on using this library to read DTCs - he probably moved on, as this issue has not seen any activity in over a year...

Regarding your question: which DTCs does the other app read, and which ones did you read using this library? There are 3 different PIDs for reading them:
- `LTOBD2PID_PERMANENT_DTC_0A`
- `LTOBD2PID_PENDING_DTC_07`
- `LTOBD2PID_STORED_DTC_03`

**Comment by mukulpandey57** [comment]
@thdankert , Thanks for your quick response. I was using all three PID ($03, $07,$0A) and only getting the C Codes and other wired scanner and apps  getting the P Codes. Please assist. 

Seems like something related to this  https://github.com/mickeyl/LTSupportAutomotive/issues/31

![Screenshot 2020-08-24 at 2 47 07 PM](https://user-images.githubusercontent.com/8499425/91026734-c829af00-e618-11ea-97a8-1e0fb6f6f007.png)
![Screenshot 2020-08-24 at 2 46 57 PM](https://user-images.githubusercontent.com/8499425/91026750-cc55cc80-e618-11ea-827d-3ba1408a20e6.png)
![Screenshot 2020-08-24 at 2 46 51 PM](https://user-images.githubusercontent.com/8499425/91026766-d37cda80-e618-11ea-86a6-90486ddabb07.png)
![Screenshot 2020-08-24 at 2 46 35 PM](https://user-images.githubusercontent.com/8499425/91026700-bcd68380-e618-11ea-8615-41a49214013d.png)
![Screenshot 2020-08-24 at 2 46 44 PM](https://user-images.githubusercontent.com/8499425/91026671-b5af7580-e618-11ea-8b06-3c93c88c81e5.png)

**Comment by thdankert** [comment]
I'm afraid I am unable to assist you there. I don't know anything about your car, the code you wrote, and how the other applications read DTCs.

**Comment by Nikhil-Quadrish** [comment]
How to clear the DTC code ? "LTOBD2PID_CLEAR_STORED_DTC_04" any more explanation to this?
